modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -846,52 +846,127 @@ def join(
         HdkOnNativeDataframe
             The new frame.
         """
-        how = JoinType(how)
+        check_join_supported(how)
         assert (
             left_on is not None and right_on is not None
         ), "Merge with unspecified 'left_on' or 'right_on' parameter is not supported in the engine"
         assert len(left_on) == len(
             right_on
         ), "'left_on' and 'right_on' lengths don't match"
 
-        for col in left_on:
-            assert col in self.columns, f"'left_on' references unknown column {col}"
-        for col in right_on:
-            assert col in other.columns, f"'right_on' references unknown column {col}"
-
-        new_columns = []
+        def validate(what, df, col_names):
+            cols = df.columns
+            new_col_names = col_names
+            for i, col in enumerate(col_names):
+                if col not in cols:
+                    new_name = None
+                    if df._index_cache is not None:
+                        for j, name in enumerate(df._index_cache):
+                            if name == col:
+                                new_name = f"__index__{j}_{name}"
```

Comment:
can we create a method for this? I mean a method that would mangle index names both before putting the index columns into arrow/omnisci and would also map original names to the mangled ones, so the logic of such renaming wouldn't be all across the class but in a single place

Reply:
we can actually modify an existing method something like:
```python
def _mangle_index_names(self, original_names, index=None):
    if index is None:
        return [
            f"__index__{i}_{'__None__' if n is None else n}"
            for i, n in enumerate(original_names)
        ]
    else:
        return [
            f"__index__{index.names.index(n)}_{'__None__' if n is None else n}"
            for n in original_names
        ]

# and then use it here like:
new_col_names = self._mangle_index_names(col_names, self._index_cache)
```

Reply:
Replaced this code with the code suggested below.